### PR TITLE
EarlyStopping: restore model weights corresponding to best value of monitored quantity

### PR DESCRIPTION
**Pull request by MarcoAndreaBuchmann**
### Summary

The current implementation of EarlyStopping monitors a user-specified quantity, and stops training after [patience] epochs have passed since the last best value of the monitored quantity.

In the **current** implementation, the final model weights are those at the end of the patience period.

In the **proposed** implementation, the final model weights are those from the epoch with the best value of the monitored quantity, in agreement with the original paper [1] as well as [2].

The proposed implementation includes an additional test case (test_EarlyStopping_final_weights) in test_callbacks.py; it also makes two existing test cases (test_EarlyStopping_patience and test_EarlyStopping_baseline) compatible with the proposed implementation.

[1]
Advances in Neural Information Processing Systems 2", San Mateo, California, 1990, pages 630-637
[2]
Prechelt L. (2012) Early Stopping — But When?. In: Montavon G., Orr G.B., Müller KR. (eds) Neural Networks: Tricks of the Trade. Lecture Notes in Computer Science, vol 7700. Springer, Berlin, Heidelberg

### Related Issues
None

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [ ] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by taehoonlee** [comment]
Thank you for the PR, @MarcoAndreaBuchmann. The motivation is reasonable, however in my opinion, there might be no significant difference between N-th epoch and (N-1)-th epoch. If so, this PR suffers from computational overheads due to the `get_weights()` on every epoch.

**Comment by MarcoAndreaBuchmann** [comment]
Thanks for the feedback, @taehoonlee. The motivation for the pull request is not so much for the `N` vs `(N-1)` case, but rather when patience `p > 1`. In that case, with the current implementation, model weights for the last configuration are used, so if we call that `N`, the correct weights would be those corresponding to `N-p`, and we're off by `p` epochs. This can make a big difference if the patience parameter `p` is large, but also if the model overfits quickly. I would therefore argue that the (small) computational overhead is definitely worth it.

**Comment by taehoonlee** [comment]
@MarcoAndreaBuchmann, You are right. As `patience` grows up, differences will be bigger. However, nobody knows when the generalization error will be minimized. While you assume that users want to capture the weights on `N-p`-th epoch, we don't know the exact local optimum in reality. We can only just guess that it will lie on somewhere between the `N`-th and the `N-p`-th.

Thus, how about adding a parameter `rollback` with the default `False` and making the operations as optional when `rollback=True`?

**Comment by MarcoAndreaBuchmann** [comment]
@taehoonlee, thanks a lot for the quick reaction and your feedback!

I agree that we do not know the exact weights that generalize best. The assumption behind early stopping is that by monitoring a quantity on the validation set, we can improve generalization, so it would be consistent to assume that model weights corresponding to epoch `N-p` (where the quantity monitored on the validation set has its best value) are more likely to generalize than those corresponding to epoch `N`, where the monitored quantity on the validation set is worse.

I would like to note that I do not object to introducing an additional parameter to disable the proposed feature, e.g. `use_last_epoch_weights=True` to be super explicit, though that would change the API. I do think that the proposed implementation is more consistent with the assumptions of early stopping than the current one, though. So should we change the API, and introduce the `use_last_epoch_weights` parameter?

**Comment by MarcoAndreaBuchmann** [comment]
@taehoonlee : could you please confirm whether you're ok with the pull request (or if you'd prefer to use the `use_last_epoch_weights` parameter as mentioned above)?

**Comment by taehoonlee** [comment]
@MarcoAndreaBuchmann, I think that both the current version and this PR fit the principle of early stopping because validation errors are always fluctuating and it is hard to tell exactly when to stop training. And the existing callback can cover this PR. In other words,

- Users, who want to restore the `N-p`-th weights, can use `EarlyStopping(patience=p)` with `ModelCheckpoint(save_best_only=True)`,
- Users, who want to exploit `N`-th weights, will just use `EarlyStopping(patience=p)`.

**Comment by MarcoAndreaBuchmann** [comment]
@taehoonlee , I agree that one can achieve the result of the proposed feature using `ModelCheckpoint(save_best_only=True)`. Note, though, that if the monitored quantity on the validation set decreases for e.g. the first hundred epochs, then this means that the weights are written to disk a hundred times. Therefore, using `ModelCheckpoint(save_best_only=True)` would add much more overhead than this PR. Aside from the overhead, the original publications explainng early stopping note that the correct implementation should return the model weights with the best value on the validation set, not the last, but anyway - we have an active community here, so why not offer users both options to do early stopping?

**Comment by taehoonlee** [comment]
@MarcoAndreaBuchmann, Maybe we need @fchollet's comments.

**Comment by datumbox** [comment]
Sounds like a good idea. I hope this PR goes through. If it does it might be worth patching also `ReduceLROnPlateau` to follow the same logic.

File: keras/callbacks.py [hunk 1]
```diff
@@ -456,7 +456,9 @@ def on_epoch_end(self, epoch, logs=None):
 
 
 class EarlyStopping(Callback):
-    """Stop training when a monitored quantity has stopped improving.
+    """Stop training when a monitored quantity has stopped improving and
```
Comment (fchollet) [inline keras/callbacks.py]: Please standardize the docstring; the docstring summary (top line) should start with a single line ending in a dot. More information can be added after a line break.

File: keras/callbacks.py [hunk 2]
```diff
@@ -495,6 +497,7 @@ def __init__(self,
         self.min_delta = min_delta
         self.wait = 0
         self.stopped_epoch = 0
+        self.good_weights = None
```
Comment (fchollet) [inline keras/callbacks.py]: This should be `best_weights`

File: keras/callbacks.py [hunk 3]
```diff
@@ -538,11 +541,13 @@ def on_epoch_end(self, epoch, logs=None):
         if self.monitor_op(current - self.min_delta, self.best):
             self.best = current
             self.wait = 0
+            self.good_weights = self.model.get_weights()
         else:
             self.wait += 1
             if self.wait >= self.patience:
                 self.stopped_epoch = epoch
                 self.model.stop_training = True
+                self.model.set_weights(self.good_weights)
```
Comment (fchollet) [inline keras/callbacks.py]: Please print a message for this action (like we do in `on_train_end`): "Restoring model to its state at the end of the best epoch."

**Comment by taehoonlee** [comment]
LGTM. Thank you for all your support, @MarcoAndreaBuchmann.

**Comment by mmalekzadeh** [comment]
Why is this "restore_best_weights" not available in the current version 2.2.2?

**Comment by MarcoAndreaBuchmann** [comment]
Hi @mmalekzadeh , thanks for the interest in this PR. The code has been merged into master about a month ago (Aug 18), but 2.2.2 had been released before that (on Jul 28), so it is not part of 2.2.2.
